File: src/main/python/AoC2024_16.py
#! /usr/bin/env python3
#
# Advent of Code 2024 Day 16
#

from __future__ import annotations

import itertools
import sys
from typing import Iterator
from typing import NamedTuple
from typing import Self

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples
from aoc.geometry import Direction
from aoc.graph import Dijkstra
from aoc.grid import Cell

IDX_TO_DIR = {
    0: Direction.UP,
    1: Direction.RIGHT,
    2: Direction.DOWN,
    3: Direction.LEFT,
}
DIR_TO_IDX = {
    Direction.UP: 0,
    Direction.RIGHT: 1,
    Direction.DOWN: 2,
    Direction.LEFT: 3,
}
TURNS = {
    Direction.UP: [Direction.LEFT, Direction.RIGHT],
    Direction.RIGHT: [Direction.UP, Direction.DOWN],
    Direction.DOWN: [Direction.LEFT, Direction.RIGHT],
    Direction.LEFT: [Direction.UP, Direction.DOWN],
}


TEST1 = """\
###############
#.......#....E#
#.#.###.#.###.#
#.....#.#...#.#
#.###.#####.#.#
#.#.#.......#.#
#.#.#####.###.#
#...........#.#
###.#.#####.#.#
#...#.....#.#.#
#.#.#.###.#.#.#
#.....#...#.#.#
#.###.#.#.#.#.#
#S..#.....#...#
###############
"""
TEST2 = """\
#################
#...#...#...#..E#
#.#.#.#.#.#.#.#.#
#.#.#.#...#...#.#
#.#.#.#.###.#.#.#
#...#.#.#.....#.#
#.#.#.#.#.#####.#
#.#...#.#.#.....#
#.#.#####.#.###.#
#.#.#.......#...#
#.#.###.#####.###
#.#.#...#.....#.#
#.#.#.#####.###.#
#.#.#.........#.#
#.#.#.#########.#
#S#.............#
#################
"""


class ReindeerMaze(NamedTuple):
    grid: list[list[str]]
    start: Cell
    end: Cell

    @classmethod
    def from_grid(cls, strings: list[str]) -> Self:
        grid = [[ch for ch in line] for line in strings]
        start = Cell(len(grid) - 2, 1)
        end = Cell(1, len(grid[0]) - 2)
        return cls(grid, start, end)

    def adjacent(self, state: State) -> Iterator[State]:
        r, c, dir = state
        direction = IDX_TO_DIR[dir]
        states = (
            (r - d.y, c + d.x, DIR_TO_IDX[d])
            for d in [direction] + TURNS[direction]
        )
        for state in states:
            if self.grid[state[0]][state[1]] != "#":
                yield state


Input = ReindeerMaze
Output1 = int
Output2 = int
State = tuple[int, int, int]


class Solution(SolutionBase[Input, Output1, Output2]):

    def parse_input(self, input_data: InputData) -> Input:
        return ReindeerMaze.from_grid(list(input_data))

    def part_1(self, maze: Input) -> Output1:
        return Dijkstra.distance(
            (maze.start.row, maze.start.col, DIR_TO_IDX[Direction.RIGHT]),
            lambda state: (state[0], state[1]) == (maze.end.row, maze.end.col),
            lambda state: maze.adjacent(state),
            lambda curr, nxt: 1 if curr[2] == nxt[2] else 1001,
        )

    def part_2(self, maze: Input) -> Output2:
        result = Dijkstra.all(
            (maze.start.row, maze.start.col, DIR_TO_IDX[Direction.RIGHT]),
            lambda state: (state[0], state[1]) == (maze.end.row, maze.end.col),
            lambda state: maze.adjacent(state),
            lambda curr, nxt: 1 if curr[2] == nxt[2] else 1001,
        )
        return len(
            {
                (state[0], state[1])
                for end, dir in itertools.product(
                    [maze.end], Direction.capitals()
                )
                for paths in result.get_paths(
                    (end.row, end.col, DIR_TO_IDX[dir])
                )
                for state in paths
            }
        )

    @aoc_samples(
        (
            ("part_1", TEST1, 7036),
            ("part_1", TEST2, 11048),
            ("part_2", TEST1, 45),
            ("part_2", TEST2, 64),
        )
    )
    def samples(self) -> None:
        pass


solution = Solution(2024, 16)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()

File: src/main/java/AoC2024_16.java
import static com.github.pareronia.aoc.IterTools.product;

import java.util.List;
import java.util.Set;
import java.util.stream.Stream;

import com.github.pareronia.aoc.Grid.Cell;
import com.github.pareronia.aoc.geometry.Direction;
import com.github.pareronia.aoc.geometry.Turn;
import com.github.pareronia.aoc.graph.Dijkstra;
import com.github.pareronia.aoc.graph.Dijkstra.AllResults;
import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

public final class AoC2024_16
        extends SolutionBase<AoC2024_16.ReindeerMaze, Integer, Integer> {
    
    private AoC2024_16(final boolean debug) {
        super(debug);
    }
    
    public static AoC2024_16 create() {
        return new AoC2024_16(false);
    }
    
    public static AoC2024_16 createDebug() {
        return new AoC2024_16(true);
    }
    
    @Override
    protected ReindeerMaze parseInput(final List<String> inputs) {
        return ReindeerMaze.fromInput(inputs);
    }

    @Override
    public Integer solvePart1(final ReindeerMaze maze) {
        return (int) Dijkstra.distance(
                new State(maze.start, Direction.RIGHT),
                state -> state.cell.equals(maze.end),
                state -> maze.adjacent(state),
                (curr, next) -> curr.direction == next.direction ? 1 : 1001);
    }
    
    @Override
    public Integer solvePart2(final ReindeerMaze maze) {
        final AllResults<State> result = Dijkstra.all(
                new State(maze.start, Direction.RIGHT),
                state -> state.cell.equals(maze.end),
                state -> maze.adjacent(state),
                (curr, next) -> curr.direction == next.direction ? 1 : 1001);
        return (int) product(List.of(maze.end), Direction.CAPITAL).stream()
            .flatMap(pp -> result.getPaths(new State(pp.first(), pp.second())).stream())
            .flatMap(List::stream)
            .map(State::cell)
            .distinct()
            .count();
    }
    
    @Override
    @Samples({
        @Sample(method = "part1", input = TEST1, expected = "7036"),
        @Sample(method = "part1", input = TEST2, expected = "11048"),
        @Sample(method = "part2", input = TEST1, expected = "45"),
        @Sample(method = "part2", input = TEST2, expected = "64"),
    })
    public void samples() {
    }
    
    public static void main(final String[] args) throws Exception {
        AoC2024_16.create().run();
    }

    private static final String TEST1 = """
            ###############
            #.......#....E#
            #.#.###.#.###.#
            #.....#.#...#.#
            #.###.#####.#.#
            #.#.#.......#.#
            #.#.#####.###.#
            #...........#.#
            ###.#.#####.#.#
            #...#.....#.#.#
            #.#.#.###.#.#.#
            #.....#...#.#.#
            #.###.#.#.#.#.#
            #S..#.....#...#
            ###############
            """;
    private static final String TEST2 = """
            #################
            #...#...#...#..E#
            #.#.#.#.#.#.#.#.#
            #.#.#.#...#...#.#
            #.#.#.#.###.#.#.#
            #...#.#.#.....#.#
            #.#.#.#.#.#####.#
            #.#...#.#.#.....#
            #.#.#####.#.###.#
            #.#.#.......#...#
            #.#.###.#####.###
            #.#.#...#.....#.#
            #.#.#.#####.###.#
            #.#.#.........#.#
            #.#.#.#########.#
            #S#.............#
            #################
            """;
    
    record State(Cell cell, Direction direction) {}

    record ReindeerMaze(char[][] grid, Cell start, Cell end) {

        public static ReindeerMaze fromInput(final List<String> strings) {
            final int h = strings.size();
            final int w = strings.get(0).length();
            final char[][] cells = new char[h][w];
            for (int i = 0; i < h; i++) {
                cells[i] = strings.get(i).toCharArray();
            }
            return new ReindeerMaze(cells, Cell.at(h - 2, 1), Cell.at(1, w - 2));
        }
        
        public Stream<State> adjacent(final State state) {
            return Set.of(
                        state.direction,
                        state.direction.turn(Turn.RIGHT),
                        state.direction.turn(Turn.LEFT)
                    ).stream()
                    .map(dir -> new State(state.cell.at(dir), dir))
                    .filter(st -> grid[st.cell.getRow()][st.cell.getCol()] != '#');
        }
    }
}